noir-projects/noir-contracts/contracts/app/orderbook_contract/src/main.nr

```diff
@@ -0,0 +1,198 @@
+
+        OrderCreated { order_id }.emit(encode_event(&mut context));
+    }
+
+    /// Privately fulfills an existing order in the orderbook
+    /// The taker provides the order ID they want to fulfill
+    #[private]
+    fn fulfill_order(order_id: Field, nonce: Field) {
+        let config = storage.config.read();
+        let order = storage.orders.at(order_id).read();
+        let taker = context.msg_sender();
+
+        // Determine which tokens are being exchanged based on bid_token_is_zero flag
+        let (bid_token, ask_token) = config.get_tokens(order.bid_token_is_zero);
+
+        // TODO(#14362): Once we allow partial note completion in private we can skip the transfer to public here and
+        // just finalize the maker's partial note straight away here.
+        // Transfer the ask_amount from taker to the contract
+        Token::at(ask_token)
+            .transfer_to_public(taker, context.this_address(), order.ask_amount, nonce)
+            .call(&mut context);
+
+        // Prepare partial note for taker to receive bid_token
+        let taker_partial_note =
+            Token::at(bid_token).prepare_private_balance_increase(taker, taker).call(&mut context);
+
+        // Nullify the order such that it cannot be fulfilled again. We emit a nullifier instead of deleting the order
+        // from public storage because we get no refund for resetting public storage to zero and just emitting
+        // a nullifier is cheaper (1 Field in DA instead of multiple Fields for the order). We use the `order_id`
+        // itself as the nullifier because this contract does not work with notes and hence there is no risk of
+        // colliding with a real note nullifier.
+        context.push_nullifier(order_id);
+
+        // Enqueue the fulfillment to finalize both partial notes
+        Orderbook::at(context.this_address())
+            ._fulfill_order(
+                order_id,
+                taker_partial_note,
+                bid_token,
+                ask_token,
+                order.bid_amount,
+                order.ask_amount,
+            )
+            .enqueue(&mut context);
+    }
+
+    #[public]
+    #[internal]
+    fn _fulfill_order(
+        order_id: Field,
+        taker_partial_note: PartialUintNote,
+        bid_token: AztecAddress,
+        ask_token: AztecAddress,
+        bid_amount: u128,
+        ask_amount: u128,
+    ) {
+        // The `order_id` is a serialized form of the maker's partial note.
+        let maker_partial_note = PartialUintNote::from_field(order_id);
+
+        // Finalize transfer of ask_amount of ask_token to maker
+        Token::at(ask_token).finalize_transfer_to_private(ask_amount, maker_partial_note).call(
+            &mut context,
+        );
+
+        // Finalize transfer of bid_amount of bid_token to taker
+        Token::at(bid_token).finalize_transfer_to_private(bid_amount, taker_partial_note).call(
+            &mut context,
```

Comment:
by leveraging `transfer_public_to_private + transfer_private_to_commitment` we can also hide what's the ultimate ask & bid amounts, enabling other fulfillment mechanisms (in comparison with fill-or-kill as in this one)

Reply:
Hello 👋 
goal of this PR is to have a way to test [partial notes completion in private](https://github.com/AztecProtocol/aztec-packages/pull/14533) and not to create a real Orderbook DEX.